### PR TITLE
[Frontend] Expose custom args in OpenAI APIs

### LINKED ISSUE #16802: [Feature]: Support custom args in OpenAI (chat) completion requests
https://github.com/vllm-project/vllm/pull/13300 added an `extra_args` field to `SamplingParameters` to allow for passing arbitrary parameters through to custom sampler / logits processor implementations.

A follow on is to expose this in the OpenAI API endpoints, see discussion at the end of that PR.

This will be needed to support custom sampling parameters via V1 logits processor plugins once that's ready (https://github.com/vllm-project/vllm/pull/13360).

### PR DESCRIPTION
**Pull request by afeldman-nm**
Add a `vllm_xargs: Optional[dict[str, Union[str,int,float]]]` field to `CompletionRequest`, `ChatCompletionRequest` and `TranscriptionRequest` (these are the only OpenAIBaseModel subclasses which had a `logits_processors` field in v0.) This field is injected into `SamplingParams.extra_args` via `SamplingParams.from_optional()`; each dict key/value pair in `extra_args` becomes an assignment to an attribute of `sampling_params`.

## Purpose

Enable extensible features such as logits processors and plugins to receive arbitrary custom arguments via the REST API. Mirror `SamplingParams.extra_args` in the REST API.

## Test plan

Does not require additional unit tests (when logitsprocs extensibility is introduced later, this will implicitly test custom args). Pre-existing unit tests must pass so we know pre-existing features are not being broken.

## Test results

N/A

## Documentation changes

* `SamplingParams` docstring clarifies that `extra_args` may plumb custom args to logitsprocs, plugins, etc. (previous it just said logitsprocs)
* In `vllm/entrypoints/openai/protocol.py`: for `CompletionRequest`, `ChatCompletionRequest`, and `TranscriptionRequest`, move `vllm_xargs` definition inside the `# --8<-- [start:completion-extra-params]` section and add more detail to the description string.

## Final note

The pre-existing behavior of `protocol.py` `ChatCompletionRequest` and `CompletionRequest` is that `kv_transfer_params` is passed into the engine via `SamplingParams.extra_args`; this PR simply merges `vllm_xargs` into `SamplingParams.extra_args` alongside `kv_transfer_params`. In the future it may be worth considering whether `SamplingParams.extra_args` is the best pathway for plumbing `kv_transfer_params` into the engine; it would seem to break the convention that `SamplingParams.extra_args` is not intended for "in-tree" functionality.

RFC: #17191 

Fixes #16802

### CODE REVIEW

**Comment by njhill** [comment]
Thanks @afeldman-nm! It would be good to include a test that shows how these can be passed via the OpenAI client sdk using its `extra_body` option: https://github.com/openai/openai-python?tab=readme-ov-file#undocumented-request-params

I'm unsure whether we want these new custom args to be in a nested json object (as you've done here) or just extra top-level args.

**Comment by afeldman-nm** [comment]
Thanks @njhill . Agree regarding the unit test. I need to think a bit about the right way to do it

**Comment by afeldman-nm** [comment]
@njhill @comaniac @WoosukKwon 

CC: @robertgshaw2-redhat

File: tests/v1/entrypoints/openai/test_completion.py [hunk 1]
```diff
@@ -80,6 +80,66 @@ async def test_single_completion(client: openai.AsyncOpenAI,
...
+    in `extra_sampling_params`; test that the value is applied.
+    """
+    completion_body = await client.completions.create(
+        model=model_name,
+        prompt="Hello, my name is",
+        max_tokens=10,
+        temperature=0.0,
+        # Contradictory `max_tokens`
+        extra_body={
+            "max_tokens": 5,
```
Comment (comaniac) [inline tests/v1/entrypoints/openai/test_completion.py:104]: I don't think we should allow conflict fields in extra body as it is very confusing. The "extra_xxx" should literally extra fields that are not defined in other places.

File: tests/v1/entrypoints/openai/test_completion.py [hunk 2]
```diff
@@ -80,6 +80,66 @@ async def test_single_completion(client: openai.AsyncOpenAI,
...
+        model=model_name,
+        prompt="Hello, my name is",
+        temperature=0.0,
+        # Contradictory `max_tokens`
+        extra_body={
+            "ignore_eos": True,
+            "extra_sampling_params": {
+                # Contradictory max_tokens
+                "max_tokens": 5
+            }
```
Comment (comaniac) [inline tests/v1/entrypoints/openai/test_completion.py:128]: Intuitively ingore_eos is a part of extra_sampling_params so we may consider moving it in? But this may affect existing users we should be careful. For example we should allow it in both places and show deprecated warning for 1-2 releases. Open to discuss cc @WoosukKwon @njhill

File: vllm/entrypoints/openai/protocol.py [hunk 3]
```diff
@@ -242,6 +242,12 @@ class ChatCompletionRequest(OpenAIBaseModel):
         ChatCompletionNamedToolChoiceParam,
     ]] = "none"
 
+    # Custom sampling params
+    extra_sampling_params: Optional[dict[str, Any]] = Field(
+        default=None,
+        description=("Additional kwargs to pass to sampling."),
```
Comment (comaniac) [inline vllm/entrypoints/openai/protocol.py:248]: Better to have more details, such as where to see the doc about extra params

File: vllm/sampling_params.py [hunk 4]
```diff
@@ -242,7 +242,6 @@ class SamplingParams(
     guided_decoding: Optional[GuidedDecodingParams] = None
     logit_bias: Optional[dict[int, float]] = None
     allowed_token_ids: Optional[list[int]] = None
-    extra_args: Optional[dict[str, Any]] = None
```
Comment (comaniac) [inline vllm/sampling_params.py:245]: Why we need to change this? I suppose we only need to change the frontend to take extra args from OpenAI protocol, but the underlying implementation could remain the same?

**Comment by afeldman-nm** [comment]
Thanks for your review @comaniac . After chatting with Cody, I think this interface change is sufficiently impactful to merit an RFC which I will write and share shortly.

File: tests/v1/entrypoints/openai/test_completion.py (hunk 1 above)
Comment (afeldman-nm) [reply tests/v1/entrypoints/openai/test_completion.py:104]: PSA after chatting with Cody: `extra_body` is entirely a "client-side" feature, i.e. the Python SDK takes an `extra_body` argument, extracts the key/value pairs, and injects them into the JSON request as if they were top-level arguments. This means that the server never sees `extra_body` and in fact does not have the ability to handle it (If you use the HTTP client, any arguments in `extra_body` get ignored.) Another implication of this, is that we have no control over the behavior of `extra_body` because it is part of OpenAI's client SDK. And as you can see here:

https://github.com/openai/openai-python/blob/ed53107e10e6c86754866b48f8bd862659134ca8/src/openai/resources/models.py#L48

"The extra values given here {in `extra_body`} take precedence over values defined on the client or passed to this method."

So the support for having conflicting settings in `extra_body` is not a new feature added by my PR, it is an immutable (to us) characteristic of the SDK. I am simply utilizing this feature as a testing hack to confirm that `extra_body` is working.

File: tests/v1/entrypoints/openai/test_completion.py (hunk 2 above)
Comment (afeldman-nm) [reply tests/v1/entrypoints/openai/test_completion.py:128]: I'm going to write a short RFC for custom args because I think these details need to be ironed out. In the near-term I think it is fine to allow `special` fields to be set in two ways: as top-level API arguments (current behavior) or via `extra_sampling_params` (new behavior in this PR.)

In the long term, having two ways to set a special arg such as `ignore_eos` is confusing & we should probably restrict special args to only be set via `extra_sampling_params`. However, it is likely that customers depend on the current behavior & their code would be broken by this change. So we should probably wait until a major release (in whatever appropriate sense of the word "major") before restricting special args to be solely configurable via `extra_sampling_params`.

File: vllm/entrypoints/openai/protocol.py (hunk 3 above)
Comment (afeldman-nm) [reply vllm/entrypoints/openai/protocol.py:248]: yes I will clarify. I should probably also add some docs changes (perhaps as a separate PR)

**Comment by helloworld1** [comment]
@afeldman-nm Are there any update on this PR? Have we reached an agreement in the RFC? Thanks!

**Comment by afeldman-nm** [comment]
> @afeldman-nm Are there any update on this PR? Have we reached an agreement in the RFC? Thanks!

Hello @helloworld1 , yes the final proposal can be found here: https://github.com/vllm-project/vllm/issues/17191#issuecomment-2858443302

Although the RFC is finalized, this PR is has stalled because it is a bit tricky to unit test custom args when the feature it was designed to support (custom logits processors #17799  based on new logits processor implementation #16728 ) is itself still in development. However the custom args workstream is  still WIP and is not abandoned.

**Comment by helloworld1** [comment]
@afeldman-nm Glad to see it is still in progress. My use case is passing in `truncate_prompt_tokens` sampling parameters. Can we just unit test what we can test now and add more comprehensive unit tests when the logits process work is done?

**Comment by afeldman-nm** [comment]
> @afeldman-nm Glad to see it is still in progress. My use case is passing in `truncate_prompt_tokens` sampling parameters. Can we just unit test what we can test now and add more comprehensive unit tests when the logits process work is done?

Hi @helloworld1 - working on getting this PR landed as-is

File: vllm/entrypoints/openai/protocol.py [hunk 5]
```diff
@@ -252,6 +252,12 @@ class ChatCompletionRequest(OpenAIBaseModel):
         ChatCompletionNamedToolChoiceParam,
     ]] = "none"
 
+    # Custom sampling params
+    vllm_xargs: Optional[dict[str, Union[str, int, float]]] = Field(
+        default=None,
+        description=("Additional kwargs to pass to sampling."),
```
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py]: Suggest rewording: `"Additional parameters for custom extensions"

Since this is documenting the external HTTP API, it probably doesn't make sense to refer to them as `kwargs`

(same for the other places)

File: vllm/entrypoints/openai/protocol.py [hunk 6]
```diff
@@ -553,8 +559,12 @@ def to_sampling_params(
             logit_bias=self.logit_bias,
             bad_words= self.bad_words,
             allowed_token_ids=self.allowed_token_ids,
-            extra_args=({"kv_transfer_params": self.kv_transfer_params}
-                        if self.kv_transfer_params else None))
+            # Pass in `extra_body` args and kv transfer params
+            extra_args={**({"kv_transfer_params": self.kv_transfer_params}
+                            if self.kv_transfer_params else {}),
+                            **(self.vllm_xargs if self.vllm_xargs else {})
+                         }
```
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py]: Would be better to move this logic to a variable declared separately above I think. Something like:
```python
extra_args = self.vllm_xargs if self.vllm_xargs is not None else {}
if self.kv_transfer_params:
    extra_args["kv_transfer_params"] = self.kv_transfer_params
extra_args = extra_args or None
```

**Review by njhill — COMMENTED** [review]
Thanks @afeldman-nm LGTM, just a couple of minor comments.

File: vllm/entrypoints/openai/protocol.py (hunk 5 above)
Comment (afeldman-nm) [reply vllm/entrypoints/openai/protocol.py]: Agreed. Here is the new more detailed description, without kwargs

```
"Dict of arbitrary additional keyword args with string or numeric "
            "values. Can be used by custom sampling implementations, plugins, "
            "etc. Not used by any in-tree sampling implementations."
```

I realized actually quite a few descriptions in `protocol.py` used the word kwargs :) I went through and changed them all to "keyword args".

File: vllm/entrypoints/openai/protocol.py (hunk 6 above)
Comment (afeldman-nm) [reply vllm/entrypoints/openai/protocol.py]: Done.

File: vllm/sampling_params.py (hunk 4 above)
Comment (afeldman-nm) [reply vllm/sampling_params.py]: Update: this change was reverted.

File: vllm/entrypoints/openai/protocol.py (hunk 3 above)
Comment (afeldman-nm) [reply vllm/entrypoints/openai/protocol.py]: Added detail and improved wording.

File: tests/v1/entrypoints/openai/test_completion.py (hunk 2 above)
Comment (afeldman-nm) [reply tests/v1/entrypoints/openai/test_completion.py]: Update: removed this unit test entirely. It is hard to write a unit test for custom args when the features that will use custom args have not yet landed. Instead this PR is being landed without unit tests; it is sufficient to show that custom args does not break any existing features. Later, when the ability to add custom logits processors is added to v1, this will create opportunities to implicitly test the custom args feature in the course of end-to-end testing of custom logits processors.

File: vllm/entrypoints/openai/protocol.py [hunk 7]
```diff
@@ -326,12 +326,13 @@ class ChatCompletionRequest(OpenAIBaseModel):
...
-        description=("Additional kwargs to pass to the template renderer. "
-                     "Will be accessible by the chat template."),
+        description=(
+            "Additional keyword args to pass to the template renderer. "
+            "Will be accessible by the chat template."),
     )
     mm_processor_kwargs: Optional[dict[str, Any]] = Field(
         default=None,
-        description=("Additional kwargs to pass to the HF processor."),
+        description=("Additional keyword args to pass to the HF processor."),
```
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py]: Actually I think kwargs is more suitable here since it's explicitly talking about what to pass to these python components, maybe leave it as is here?

File: tests/v1/entrypoints/openai/test_completion.py (hunk 1 above)
Comment (afeldman-nm) [reply tests/v1/entrypoints/openai/test_completion.py]: Update: removed this unit test anyway. But FYI to set custom args via OpenAI SDK you would pass `vllm_xargs` to `extra_body`. The comments in `protocol.py` for `vllm_xargs` have been update, and once the docs are recompiled, this is should be clear to someone reading [this page](https://docs.vllm.ai/en/v0.4.1/serving/openai_compatible_server.html#extra-parameters) that if they want to set `vllm_xargs` through the OpenAI SDK they need to use `extra_body`.

File: vllm/entrypoints/openai/protocol.py [hunk 8]
```diff
@@ -414,6 +415,14 @@ class ChatCompletionRequest(OpenAIBaseModel):
         default=None,
         description="KVTransfer parameters used for disaggregated serving.")
 
+    vllm_xargs: Optional[dict[str, Union[str, int, float]]] = Field(
+        default=None,
+        description=(
+            "Dict of arbitrary additional keyword args with string or numeric "
+            "values. Can be used by custom sampling implementations, plugins, "
+            "etc. Not used by any in-tree sampling implementations."),
```
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py]: Suggest simplification
```suggestion
            "Additional request parameters with string or "
            "numeric values, used by custom extensions."),
```

File: vllm/entrypoints/openai/protocol.py (hunk 7 above)
Comment (afeldman-nm) [reply vllm/entrypoints/openai/protocol.py]: Fixed all occurrences for `mm_processor_kwargs`

**Review by njhill — APPROVED** [review]
Thanks @afeldman-nm
